<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [14]:
import pandas as pd
import psycopg2
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup

In [16]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [17]:
# текст запроса
# выполняем выборку из таблицы vacancies, подсчет вакансий ведем по полю id - уникальное
query_3_1 = f'''
    select count(v.id)
    from public.vacancies as v
'''

In [18]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\TEMP\ipykernel_620\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [19]:
# текст запроса
# выполняем выборку из таблицы employers, подсчет вакансий ведем по полю id - уникальное
query_3_2 = f'''
    select count(e.id)
    from public.employers as e
'''

In [20]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\TEMP\ipykernel_620\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [21]:
# текст запроса
# выполняем выборку из таблицы areas, подсчет вакансий ведем по полю id - уникальное
query_3_3 = f'''
    select count(a.id)
    from public.areas as a
'''

In [22]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\TEMP\ipykernel_620\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [23]:
# текст запроса
# выполняем выборку из таблицы industries, подсчет вакансий ведем по полю id - уникальное
query_3_4 = f'''
    select count(i.id)
    from public.industries as i
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\TEMP\ipykernel_620\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

## Выводы по предварительному анализу данных

В текущей версии базы данных представлено 49197 вакансий. Вакансии предоставляются 23501 компаниями и покрывают собой 294 сферы деятельности. Представленные компании охватывают 1362 региона.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [25]:
# текст запроса
# выполняем выборку из таблицы полученной внутренним слиянием areas с vacancies по ключам id и  area_id,
# подсчет вакансий ведем по полю id - уникальное
query_4_1 = f'''
    select 
        a.name as area,
        count(v.id) as cnt
    from public.areas as a
    join public.vacancies as v on a.id = v.area_id
    group by a.name
    order by cnt desc
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\TEMP\ipykernel_620\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [27]:
# текст запроса
# выполняем выборку из таблицы vacancies по условию:
# одно из полей salary_from или salary_to не пустое
query_4_2 = f'''
    select 
        count(v.id) as cnt
    from public.vacancies as v
    where (v.salary_from is not Null) or (v.salary_to is not Null)    
'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\TEMP\ipykernel_620\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [29]:
# текст запроса
# выполняем выборку из таблицы vacancies,
# подсчет среднего значения по полям salary_from, salary_to
query_4_3 = f'''
    select 
        avg(v.salary_from) as avg_salary_from,
        avg(v.salary_to) as avg_salary_to
    from public.vacancies as v        
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\TEMP\ipykernel_620\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0     71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [31]:
# текст запроса
# выполняем выборку из таблицы vacancies сгуппированной по полям schedule, employment и
# подсчет вакансий ведем по полю id - уникальное,
# результат сортируем по убыванию количества вакансий
query_4_4 = f'''
    select 
        v.schedule as schedule,
        v.employment as employment,
        count(v.id) as cnt
    from public.vacancies as v
    group by v.schedule, v.employment
    order by cnt desc
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\TEMP\ipykernel_620\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [33]:
# текст запроса
# выполняем выборку из таблицы vacancies сгуппированной по полю experience и
# подсчет вакансий ведем по полю id - уникальное,
# результат сортируем по возрастанию количества вакансий
query_4_5 = f'''
    select 
        v.experience as experience,
        count(v.id) as cnt
    from public.vacancies as v
    group by v.experience
    order by count(v.id) asc
'''

In [34]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\TEMP\ipykernel_620\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий
Из представленных данных видно, что основная масса вакансий размещены в крупных городах. Также видно, что приблизительно в половине случаев (вакансий) не указана заработная плата. Среднее значение вилки заработной платы по всем вакансиям (которые имеют эти данные) составляет 71к - 110к руб. Наиболее востребованным сочетанием типа рабочего графика и типа трудоустройства является -  Полный день-Полная занятость, на втором месте идет Удаленная работа-Полная занятость. Наиболее востребованными являются кандидаты с опытом роботы от 1 года до 3 лет (приблизительно половина вакансий), второе место занимают кандидаты с опытом роботы от 3 до 6 лет (приблизительно четверть вакансий).

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [35]:
# текст запроса
# выполняем выборку из таблицы полученной внутренним слиянием employers с vacancies по ключам id и  employer_id,
# сгуппированной по названию работодателя,
# подсчет вакансий ведем по полю id - уникальное,
# таблицу сортируем по убыванию количества вакансий.
# делаем две выборки первая и пятая строка, и объеденяем их между собой.
query_5_1 = f'''
    with employer_vacancies as
    (
    select 
        e.name as employer,
        count(v.id) as cnt
    from public.employers as e
    join public.vacancies as v on e.id = v.employer_id
    group by e.name
    order by count(v.id) desc
    )
    (
    select 
        '1-ое место по количеству вакансий' as top,
        ev.employer,
        ev.cnt
    from employer_vacancies as ev
    limit 1
    )
    union
    (
    select 
        '5-ое место по количеству вакансий' as top,
        ev.employer,
        ev.cnt
    from employer_vacancies as ev
    offset 4 limit 1
    )
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\TEMP\ipykernel_620\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


top       employer   cnt
0  1-ое место по количеству вакансий         Яндекс  1933
1  5-ое место по количеству вакансий  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [37]:
# текст запроса
# выполняем выборку из таблицы полученной левым слиянием areas с employers и vacancies по ключам id и area, area_id,
# с условием area_id не пусто,
# сгуппированной по названию региона,
# подсчет работодателей и вакансий ведем по полям id - уникальное,
# таблицу сортируем по убыванию количества работодателей.
# для показа выбираем первую строку.
query_5_2 = f'''
    select 
        a.name as area_name,
        count(distinct e.id) as cnt_employers,
        count(distinct v.id) as cnt_vacancies
    from public.areas as a
    left join public.employers as e on a.id = e.area
    left join public.vacancies as v on a.id = v.area_id
    where v.area_id is null
    group by a.name
    order by cnt_employers desc
    limit 1
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\TEMP\ipykernel_620\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area_name  cnt_employers  cnt_vacancies
0    Россия            410              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [39]:
# текст запроса
# выполняем выборку из таблицы полученной левым слиянием employers с vacancies по ключам id и employer_id,
# с условием area_id не пусто,
# сгуппированной по названию работодателя,
# подсчет регионов ведем по полю area_id,
# таблицу сортируем по убыванию количества регионов.
query_5_3 = f'''
    select 
        e.name as employer_name,
        count(distinct v.area_id) as cnt_areas
    from public.employers as e
    left join public.vacancies as v on e.id = v.employer_id
    where v.area_id is not null
    group by e.name
    order by cnt_areas desc
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\TEMP\ipykernel_620\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer_name  cnt_areas
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
14761                  UniSol          1
14762            UNISTORY LLC          1
14763                   UNIT6          1
14764     United Distribution          1
14765              UNITEL LLC          1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [41]:
# текст запроса
# выполняем выборку из таблицы полученной левым слиянием employers с employers_industries по ключам id и employer_id,
# с условием industry_id пусто.
query_5_4 = f'''
    select 
        count(e.id) as cnt_with_out_industry
    from public.employers as e
    left join public.employers_industries as ei on e.id = ei.employer_id
    where ei.industry_id is null    
'''

In [42]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\TEMP\ipykernel_620\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


cnt_with_out_industry
0                   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [43]:
# текст запроса
# выполняем выборку из таблицы полученной внутренним слиянием employers с employers_industries по ключам id и employer_id,
# сгуппированной по названию работодателя,
# с условием количество сфер деятельности industry_id равно 4,
# таблицу сортируем алфавитном порядке названий компаний.
# для показа выбираем третью строку.
query_5_5 = f'''
    select 
        e.name as name_company
    from public.employers as e
    join public.employers_industries as ei on e.id = ei.employer_id
    group by e.name
    having count(ei.industry_id) = 4
    order by name_company asc
    offset 2 limit 1
'''

In [44]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\TEMP\ipykernel_620\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name_company
0         2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [45]:
# текст запроса
# выполняем выборку из таблицы полученной внутренним слиянием employers с employers_industries и industries
# по ключам id и employer_id, industry_id и id,
# с условием название сферы деятельности name равно 'Разработка программного обеспечения',
# подсчет работодателей ведем по полю id - уникальное.
query_5_5 = f'''
    select 
        count(e.id) as cnt_company_dev
    from public.employers as e
    join public.employers_industries as ei on e.id = ei.employer_id
    join public.industries as i on ei.industry_id = i.id
    where i.name = 'Разработка программного обеспечения'
'''

In [46]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\TEMP\ipykernel_620\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


cnt_company_dev
0             3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [47]:
# код для получения списка городов-милионников
# средствами пандас считываем html-страницу
cities = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
# выбираем таблицу с городами милионниками, из нее столбец с названиями городов, в формате кортеж
cities = tuple(cities[0]['Город'])
display(cities)

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [48]:
# текст запроса
# выполняем выборку из таблицы полученной внутренним слиянием vacancies с employers и areas
# по ключам employer_id и id, area_id и id,
# с условием название компании - "Яндекс" и регион входит в список городов-милионников.
# промежуточную таблицу группируем по названию города и подсчитываем количество вакансий,
# и объеденяем с подсчетом количества вакансий без группировки с названием 'total',
# результат сортируем по возрастанию количества вакансий.
query_5_6 = f'''
    with t as
    (
    select 
        a.name,
        v.id
    from public.vacancies as v
    join public.employers as e on v.employer_id = e.id
    join public.areas as a on v.area_id = a.id
    where 
        e.name='Яндекс' and
        a.name in {cities}
    )
    select 
        t.name as city,
        count(t.id) as cnt_vacancies
    from t
    group by t.name
    
    union
    
    select 
        'total' as city,
        count(t.id) as cnt_vacancies
    from t
    order by cnt_vacancies asc
 '''

In [49]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\TEMP\ipykernel_620\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


city  cnt_vacancies
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            total            485

***

# выводы по анализу работодателей
Основную массу вакансий предоставляют большие корпорации в сферах IT, финансы и добывающая индустрия. Наибольшее количество работадателей и вакансий находится в крупных административных и индустриальных регионах (городах). Наибольшее количество регионов охатывают Яндекс, Ростелеком и Спецремонт. Чуть менее половины работадателей заявляют одну сферу деятельности, соответственно остальные являются многопрофильными. Приблизительно четверть компаний занимается разработкой програмного обеспечения. Компания Яндекс является лидером по предоставлению вакансий.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [50]:
# текст запроса
# выполняем выборку из таблицы vacancies
# с условием в названии вакансии присутствуют 'data' или 'данн',
# подсчет ведем по названию вакансии name.
query_6_1 = f'''
    select 
        count(v.name) as cnt_vacancies
    from public.vacancies as v
    where lower(v.name) like '%data%' or 
          lower(v.name) like '%данн%'
'''

In [51]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\TEMP\ipykernel_620\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt_vacancies
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

**В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [52]:
# текст запроса
# выполняем выборку из таблицы vacancies
# с условием в названии вакансии присутствуют:
# 'data scientist' или
# 'data science' или
# 'исследователь данных' или
# 'ML' (без HTML)
# 'machine learning'
# 'машинн%обучен%'
# и любое из следующих условий:
# в названии вакансии присутствует 'junior' или 
# требуемый опыт — 'Нет опыта' или
# тип трудоустройства — 'Стажировка'.
# подсчет ведем по названию вакансии name.
query_6_2 = f'''
    select 
        count(v.name) as cnt_vacancies
    from public.vacancies as v
    where 
        (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
        lower(v.name) like '%machine learning%' or 
        lower(v.name) like '%машинн%обучен%') and
        (lower(v.name) like '%junior%' or 
        lower(v.experience) like '%нет опыта%' or 
        lower(v.employment) like '%стажировка%')
'''

In [53]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\TEMP\ipykernel_620\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


cnt_vacancies
0             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

**Критерии для отнесения вакансии к DS указаны в предыдущем задании.**

In [54]:
# текст запроса
# выполняем выборку из таблицы vacancies,
# с условием в названии вакансии присутствуют:
# 'data scientist' или
# 'data science' или
# 'исследователь данных' или
# 'ML' (без HTML)
# 'machine learning'
# 'машинн%обучен%'
# и любое из следующих условий:
# в именовании навыка присутствует 'junior' или 'postgres'.
# подсчет ведем по названию вакансии name.
query_6_3 = f'''
    select 
        count(v.name) as cnt_vacancies
    from public.vacancies as v
    where 
        (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
        lower(v.name) like '%machine learning%' or 
        lower(v.name) like '%машинн%обучен%') and
        (lower(v.key_skills) like '%sql%' or 
        lower(v.key_skills) like '%postgres%')
'''

In [55]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\TEMP\ipykernel_620\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


cnt_vacancies
0            229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

**Это можно сделать помощью запроса, аналогичного предыдущему.**

In [56]:
# текст запроса
# выполняем выборку из таблицы vacancies,
# с условием в названии вакансии присутствуют:
# 'data scientist' или
# 'data science' или
# 'исследователь данных' или
# 'ML' (без HTML)
# 'machine learning'
# 'машинн%обучен%'
# и следующее условие:
# в именовании навыка присутствует 'python'.
# подсчет ведем по названию вакансии name.
query_6_4 = f'''
    select 
        count(v.name) as cnt_vacancies
    from public.vacancies as v
    where 
        (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
        lower(v.name) like '%machine learning%' or 
        lower(v.name) like '%машинн%обучен%') and
        lower(v.key_skills) like '%python%'
'''

In [57]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\TEMP\ipykernel_620\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


cnt_vacancies
0            357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [58]:
# текст запроса
# выполняем выборку из таблицы vacancies,
# с условием в названии вакансии присутствуют:
# 'data scientist' или
# 'data science' или
# 'исследователь данных' или
# 'ML' (без HTML)
# 'machine learning'
# 'машинн%обучен%'
# и следующее условие:
# поле навыка не пустое.
# подсчитываем количество разделителя (знак табуляции) в поле ключевых навыков, +1 - это количество навыков.
# из промежуточной таблицы высчитываем среднее количество навыков.
query_6_5 = f'''
    with cnt_skills as
    (
    select 
        (length(v.key_skills) - 
        length(replace(v.key_skills, chr(9), ''))) + 1 as cnt
    from public.vacancies as v
    where 
        (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        (v.name like '%ML%' and v.name not like '%HTML%') or 
        lower(v.name) like '%machine learning%' or 
        lower(v.name) like '%машинн%обучен%') and
        v.key_skills is not null
    )
    select 
        round(avg(cs.cnt), 2) as avg_cnt_skills
    from cnt_skills as cs
'''

In [59]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\TEMP\ipykernel_620\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg_cnt_skills
0            6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [60]:
# текст запроса
# выполняем выборку из таблицы vacancies,
# с условием в названии вакансии присутствуют:
# 'data scientist' или
# 'data science' или
# 'исследователь данных' или
# 'ML' (без HTML)
# 'machine learning'
# 'машинн%обучен%'
# к полученной таблице применяем условие: одно из полей зарплаты не пустое
# группируем таблицу по типу опыта,
# высчитываем среднюю зарплату исключая пустые поля.
query_6_6 = f'''
    with ds_salary as
    (
    select 
        v.experience,
        v.salary_from,
        v.salary_to
    from public.vacancies as v
    where 
        lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        (v.name like '%ML%' and v.name not like '%HTML%') or 
        lower(v.name) like '%machine learning%' or 
        lower(v.name) like '%машинн%обучен%'
    )
    select 
        ds.experience as experience,
        round(avg(COALESCE((ds.salary_from + ds.salary_to)/2, ds.salary_from, ds.salary_to)), 0) as avg_salary
    from ds_salary as ds
    where ds.salary_from is not null or ds.salary_to is not null
    group by ds.experience
'''

In [61]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\TEMP\ipykernel_620\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0       От 3 до 6 лет    243115.0
1  От 1 года до 3 лет    139675.0
2           Нет опыта     74643.0

***

# выводы по предметному анализу
Около 5% вакансий связаны с работой с данными. "Молодым" специалистам в сфере DS достается 51 вакансия, что составляет приблизительно 10% вакансий в этой сфере. В сфере DS чуть меньше чем в половине вакансий требуется знание SQL или Postgres. А чуть более чем в половине вакансий требуется знание Python. Более того, кроме указанных навыков в сфере DS в среднем требуется владение 6-7 навыками. Как видно из данных с ростом опыта растет средняя зароботная плата. Как видим из данных для случая опыт более 6 лет данных о зарплате нет (возможно на этом этапе зароботная плата договорная).

Дополнительное исследование данных.
Посмотрим на самые распространенные вакансии для DS. Выведем топ 10 по колличеству размещенных вакансий.

In [62]:
# текст запроса
# выполняем выборку из таблицы vacancies,
# с условием в названии вакансии присутствуют:
# 'data scientist' или
# 'data science' или
# 'исследователь данных' или
# 'ML' (без HTML)
# 'machine learning'
# 'машинн%обучен%'
# группируем таблицу по названию вакансии,
# подсчитываем количество вакансий,
# результат сортируем в порядке убывания количества вакансий,
# показываем первые 10 значений
query_7 = f'''
    select 
          v.name as name_vacancie,
          count(v.id) as cnt_vacancie
    from public.vacancies as v
    where 
        lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        (v.name like '%ML%' and v.name not like '%HTML%') or 
        lower(v.name) like '%machine learning%' or 
        lower(v.name) like '%машинн%обучен%' 
    group by v.name
    order by cnt_vacancie desc
    limit 10 
'''

In [63]:
# результат запроса
df = pd.read_sql_query(query_7, connection)
df

C:\TEMP\ipykernel_620\3765325489.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7, connection)


name_vacancie  cnt_vacancie
0                                     Data Scientist            55
1                              Senior Data Scientist            21
2                              C++ разработчик (QML)            10
3                  ML Engineer/ Data Scientist (WAF)             9
4  Junior Data Scientist (Валидация скоринговых м...             9
5                              Middle Data Scientist             8
6                                     Data scientist             8
7              Архитектор решений с опытом в Tiny ML             6
8                          Machine Learning Engineer             6
9          Python-разработчик в команду Data Science             5

Как видим чаще всего в названии вакансии указывают Data Scientist либо Senior Data Scientist, так же мы видим что присутствует вакансия C++ разработчик (QML) которая врядли относится к DS, что свидетельствует о том что выбор условия отношения к DS не совсем корректный :)

# Общий вывод по проекту

Если говорить про рынок вакансий, то видно что наиболее высокая зарплата у дата-сейнтистов, но войти в данную отрасль без опыта достаточно сложно поскольку для начинающих специалистов очень мало предложений, а для специалистов с опытом наоборот, и судя по зарлате имеется нехватка таких специалистов.
Так же мы видим что предложения в основном сконцентрированы в самых больших городах что негативно сказывается на регионах где количество вакансий вразы меньше, а возможность удаленной работы не всегда предусмотрена что в свою очеред вызывает отток специалистов из небольших регионов.